### ISSUE TITLE
overly strict layout validation of wgsl vertex input struct

### ISSUE THREAD

**Issue by slackers123** [ScanMountGoat/wgsl_to_wgpu#34]
I have this struct in my shader:
```rust
struct Vertex {
    @location(0) position: vec3<f32>,
    @location(1) normal: vec3<f32>,
    @location(2) tangent: vec4<f32>,
    @location(3) tex_coords: vec2<f32>,
}
```
and wgsl_to_wgpu generates these asserts:
```rust
const _: () = assert!(
    std::mem::size_of::<Vertex>() == 64,
    "size of Vertex does not match WGSL"
);
const _: () = assert!(
    memoffset::offset_of!(Vertex, position) == 0,
    "offset of Vertex.position does not match WGSL"
);
const _: () = assert!(
    memoffset::offset_of!(Vertex, normal) == 16,
    "offset of Vertex.normal does not match WGSL"
);
const _: () = assert!(
    memoffset::offset_of!(Vertex, tangent) == 32,
    "offset of Vertex.tangent does not match WGSL"
);
const _: () = assert!(
    memoffset::offset_of!(Vertex, tex_coords) == 48,
    "offset of Vertex.tex_coords does not match WGSL"
);
```
Which except for the second one is incorrect.

I looked at the code for this, and it seems like an upstream problem with naga?
It might make sense to make the validation optional in that case.

Edit: the identical code was working when using an older version of this crate.

**Comment by ScanMountGoat** [comment]
The offset and size assertions look correct. For an example of WGSL's layout rules, see https://gpuweb.github.io/gpuweb/wgsl/#example-fc0bb4df. Some types like vec3 have more strict alignment rules in WGSL than in Rust. This is also true of shading languages like GLSL. If you tried to use that Rust struct with bytemuck and pass the data to the shader, you would get some unexpected values while rendering.

The one exception to this is for vertex input structs on desktop, which allow you to specify the offsets manually and don't have strict alignment requirements. The plan is to not validate structs that are only used for vertex inputs. See the discussion on #8. For now, you can either not derive bytemuck (encase handles padding at runtime), use vec4 instead of vec3, add a manual f32 padding field, delete the assertions manually, or use an older version of wgsl_to_wgpu.

**Comment by slackers123** [comment]
Ok, I guess I don't know enough about wgsl. But it was a little jarring seeing "random" errors with a struct that has worked fine until now. Thank you for your help.

**Comment by ScanMountGoat** [comment]
I'm going to leave this open to track fixing this. The Rust struct you posted is fine as long as it's only used for vertex buffers.

**Comment by ScanMountGoat** [comment]
This is resolved by add39be @slackers123, so you should be able to just use that git revision until the next crates.io release.
